### PR TITLE
ghproxy: add token hash label to cache response metric

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Without this label on the metric it's not possible to filter cache
responses by the user making the requests.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @alvaroaleman @cjwagner

### CODE REVIEW

File: ghproxy/ghcache/ghcache.go [hunk 1]
```diff
@@ -206,6 +193,16 @@ func (u upstreamTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
 }
 
+func authHeaderHash(req *http.Request) string {
+	// get authorization header to convert to sha256
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		logrus.Warn("Couldn't retrieve 'Authorization' header, adding to unknown bucket")
+		authHeader = "unknown"
+	}
+	return fmt.Sprintf("%x", sha256.Sum256([]byte(authHeader))) // use %x to make this a utf-8 string for use as a label
```
Comment (fejta) [inline ghproxy/ghcache/ghcache.go:203]: Can we choose a computationally less expensive hash?

Comment (cjwagner) [reply ghproxy/ghcache/ghcache.go:203]: This does need to be a cryptographic hash. Is SHA256 considered slow?
We could also keep a simple cache of hash values.

Comment (stevekuznetsov) [reply ghproxy/ghcache/ghcache.go:203]: Happy to follow up with a simple in-memory cache

Comment (fejta) [reply ghproxy/ghcache/ghcache.go:204]: I do not agree that this needs to be a cryptographic hash. It needs to be a one-way function that has a low chance of collisions. If we intend to perform it on every request, it must be computationally cheap.

Comment (fejta) [reply ghproxy/ghcache/ghcache.go:204]: Although it looks like we are already using sha256? In which case this isn't a regression so is less important to fix (but should be fixed, sha256 is almost certainly a poorer choice for this than say md5 given the need)

Comment (stevekuznetsov) [reply ghproxy/ghcache/ghcache.go:204]: I think the performance hit of this is miniscule but we should see when this lands ... I have added a simple in-memory cache. Isn't md5 cryptographic (but broken?)

**Comment by stevekuznetsov** [comment]
Finally done wrestling with `bazel` ... @cjwagner @fejta @alvaroaleman

File: ghproxy/ghcache/ghcache.go (hunk 1 above)
Comment (fejta) [reply ghproxy/ghcache/ghcache.go:204]: There are scenarios where md5 is unsuitable, but it isn't obvious to me how this is one of them

**Pull request merged**